File: test/src/org/orbeon/oxf/xforms/control/XFormsControlsTest.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.control;

import org.dom4j.QName;
import org.junit.Test;
import org.orbeon.oxf.test.ResourceManagerTestBase;
import org.orbeon.oxf.xforms.control.controls.XFormsInputControl;
import org.xml.sax.helpers.AttributesImpl;

import java.util.LinkedHashMap;
import java.util.Map;

import static junit.framework.Assert.assertEquals;

public class XFormsControlsTest extends ResourceManagerTestBase {

    @Test
    public void testDiffCustomMIPsChanges() {

        final AttributesImpl attributes = new AttributesImpl();
        final XFormsSingleNodeControl control1 = new XFormsInputControl(null, null, null, "input", "input-1", null) {

            private Map<String, String> customMIPs = new LinkedHashMap<String, String>();
            {
                customMIPs.put("name1", "value1");
                customMIPs.put("name2", "value2");
                customMIPs.put("name3", "value3");
                customMIPs.put("name4", "value4");
            }

            @Override
            public Map<String, String> getCustomMIPs() {
                return customMIPs;
            }
        };

        final XFormsSingleNodeControl control2 = new XFormsInputControl(null, null, null, "input", "input-2", null) {
            private Map<String, String> customMIPs = new LinkedHashMap<String, String>();
            {
                // leave as is
                customMIPs.put("name1", "value1");
                // remove name2
                // change value
                customMIPs.put("name3", "newvalue3");
                // leave as is
                customMIPs.put("name4", "value4");
            }

            @Override
            public Map<String, String> getCustomMIPs() {
                return customMIPs;
            }
        };

        XFormsSingleNodeControl.addAjaxCustomMIPs(attributes, false, control1, control2);
        assertEquals("-name2-value2 -name3-value3 +name3-newvalue3", attributes.getValue("class"));
    }

    @Test
    public void testDiffCustomMIPsNew() {

        final AttributesImpl attributes = new AttributesImpl();

        final XFormsSingleNodeControl control2 = new XFormsInputControl(null, null, null, "input", "input-1", null) {
            private Map<String, String> customMIPs = new LinkedHashMap<String, String>();
            {
                customMIPs.put("name1", "value1");
                customMIPs.put("name2", "value2");
                customMIPs.put("name3", "value3");
                customMIPs.put("name4", "value4");
            }

            @Override
            public Map<String, String> getCustomMIPs() {
                return customMIPs;
            }
        };

        XFormsSingleNodeControl.addAjaxCustomMIPs(attributes, false, null, control2);
        assertEquals("name1-value1 name2-value2 name3-value3 name4-value4", attributes.getValue("class"));
    }

    @Test
    public void testDiffClassAVT() {
        final AttributesImpl attributes = new AttributesImpl();

        final XFormsSingleNodeControl control1 = new XFormsInputControl(null, null, null, "input", "input-1", null) {
            @Override
            public String getExtensionAttributeValue(QName attributeName) {
                return "foo bar gaga";
            }
        };

        final XFormsSingleNodeControl control2 = new XFormsInputControl(null, null, null, "input", "input-2", null) {
            @Override
            public String getExtensionAttributeValue(QName attributeName) {
                return "bar toto";
            }
        };

        XFormsSingleNodeControl.addAjaxClass(attributes, false, control1, control2);
        assertEquals("-foo -gaga +toto", attributes.getValue("class"));
    }

    @Test
    public void testDiffClassAVTNew() {
        final AttributesImpl attributes = new AttributesImpl();

        final XFormsSingleNodeControl control2 = new XFormsInputControl(null, null, null, "input", "input-1", null) {
            @Override
            public String getExtensionAttributeValue(QName attributeName) {
                return "foo bar";
            }
        };

        XFormsSingleNodeControl.addAjaxClass(attributes, false, null, control2);
        assertEquals("foo bar", attributes.getValue("class"));
    }

    // NOTE: started writing this test, but just using an XFormsOutputControl without the context of an XFormsContainingDocument seems a dead-end!
//    @Test
//    public void testOutputControlRewrite() {
//
//        final Document document = Dom4jUtils.readFromURL("oxf:/org/orbeon/oxf/xforms/processor/test-form.xml", false, false);
//        final DocumentWrapper documentWrapper = new DocumentWrapper(document, null, new Configuration());
//        final Element outputElement = (Element) ((NodeWrapper) XPathCache.evaluateSingle(new PipelineContext(), documentWrapper, "(//xhtml:body//xforms:output)[1]", XFormsDocumentAnnotatorContentHandlerTest.BASIC_NAMESPACE_MAPPINGS, null, null, null, null, null)).getUnderlyingNode();
//
//        final PipelineContext pipelineContext = new PipelineContext();
//
//        final XBLContainer container = new XBLContainer("", null) {};
//        final XFormsOutputControl control1 = new XFormsOutputControl(container, null, outputElement, "output", "output-1");
//        control1.setBindingContext(pipelineContext, new XFormsContextStack.BindingContext(null, null, Collections.singletonList(documentWrapper.wrap(outputElement)), 1, "output-1", true, outputElement, null, false, null));
//
//        control1.evaluateIfNeeded(pipelineContext);
//
//        assertEquals("", control1.getExternalValue(pipelineContext));
//    }
}

File: xforms/jvm/src/test/scala/org/orbeon/oxf/xforms/control/XFormsControlsTest.scala
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.control

import org.mockito.{Matchers, Mockito}
import org.orbeon.dom.QName
import org.orbeon.oxf.test.{DocumentTestBase, ResourceManagerSupport}
import org.orbeon.oxf.xforms.analysis.{ElementAnalysis, PartAnalysis}
import org.orbeon.oxf.xforms.control.controls.XFormsInputControl
import org.orbeon.oxf.xforms.XFormsContainingDocument
import org.scalatest.funspec.AnyFunSpecLike
import org.xml.sax.helpers.AttributesImpl

class XFormsControlsTest
  extends DocumentTestBase
     with ResourceManagerSupport
     with AnyFunSpecLike {

  // Mock just what's needed to make XFormsInputControl as used below happy
  private def getContainingDocument(id: String): XFormsContainingDocument = {

    val doc = Mockito.mock(classOf[XFormsContainingDocument])
    Mockito.when(doc.getContainingDocument).thenReturn(doc)

    val elementAnalysis = Mockito.mock(classOf[ElementAnalysis])
    Mockito.when(elementAnalysis.staticId).thenReturn(id)
    Mockito.when(elementAnalysis.prefixedId).thenReturn(id)

    val partAnalysis = Mockito.mock(classOf[PartAnalysis])
    Mockito.when(partAnalysis.getControlAnalysis(Matchers.anyString)).thenReturn(elementAnalysis)
    Mockito.when(doc.getPartAnalysis).thenReturn(partAnalysis)

    doc
  }

  describe("Diff custom MIPs changes ") {
    val attributes = new AttributesImpl
    val control1 = new XFormsInputControl(getContainingDocument("input-1"), null, null, "input-1") {
      override val customMIPs = Map(
        "name1" -> "value1",
        "name2" -> "value2",
        "name3" -> "value3",
        "name4" -> "value4"
      )
    }
    val control2 = new XFormsInputControl(getContainingDocument("input-2"), null, null, "input-2") {
      override val customMIPs = Map(
        // leave as is
        "name1" -> "value1",
        // remove name2
        // change value
        "name3" -> "newvalue3",
        // leave as is
        "name4" -> "value4"
      )
    }
    XFormsSingleNodeControl.addAjaxCustomMIPs(attributes, Some(control1), control2)
    it ("must report the correct attribute changes") {
      assert("-name2-value2 -name3-value3 +name3-newvalue3" === attributes.getValue("class"))
    }
  }

  describe("Diff new custom MIPs") {
    val attributes = new AttributesImpl
    val control2 = new XFormsInputControl(getContainingDocument("input-1"), null, null, "input-1") {
      override val customMIPs = Map(
        "name1" -> "value1",
        "name2" -> "value2",
        "name3" -> "value3",
        "name4" -> "value4"
      )
    }
    XFormsSingleNodeControl.addAjaxCustomMIPs(attributes, None, control2)
    it ("must report the correct attribute changes") {
      assert("+name1-value1 +name2-value2 +name3-value3 +name4-value4" === attributes.getValue("class"))
    }
  }

  describe("Diff class AVT changes") {
    val attributes = new AttributesImpl
    val control1 = new XFormsInputControl(getContainingDocument("input-1"), null, null, "input-1") {
      override def extensionAttributeValue(attributeName: QName) = Some("foo bar gaga")
    }
    val control2 = new XFormsInputControl(getContainingDocument("input-2"), null, null, "input-2") {
      override def extensionAttributeValue(attributeName: QName) = Some("bar toto")
    }
    ControlAjaxSupport.addAjaxClasses(attributes, Some(control1), control2)
    it ("must report the correct attribute changes") {
      assert("-foo -gaga +toto" === attributes.getValue("class"))
    }
  }

  describe("Diff new class AVT") {
    val attributes = new AttributesImpl
    val control2 = new XFormsInputControl(getContainingDocument("input-1"), null, null, "input-1") {
      override def extensionAttributeValue(attributeName: QName) = Some("foo bar")
    }
    ControlAjaxSupport.addAjaxClasses(attributes, None, control2)
    it ("must report the correct attribute changes") {
      assert("foo bar" === attributes.getValue("class"))
    }
  }

  // NOTE: started writing this test, but just using an XFormsOutputControl without the context of an XFormsContainingDocument seems a dead-end!
//    @Test
//    public void testOutputControlRewrite() {
//
//        final Document document = Dom4jUtils.readFromURL("oxf:/org/orbeon/oxf/xforms/processor/test-form.xml", false, false);
//        final DocumentWrapper documentWrapper = new DocumentWrapper(document, null, new Configuration());
//        final Element outputElement = (Element) ((NodeWrapper) XPathCache.evaluateSingle(new PipelineContext(), documentWrapper, "(//xh:body//xf:output)[1]", XFormsDocumentAnnotatorContentHandlerTest.BASIC_NAMESPACE_MAPPINGS, null, null, null, null, null)).getUnderlyingNode();
//
//        final PipelineContext pipelineContext = new PipelineContext();
//
//        final XBLContainer container = new XBLContainer("", null) {};
//        final XFormsOutputControl control1 = new XFormsOutputControl(container, null, outputElement, "output", "output-1");
//        control1.setBindingContext(pipelineContext, new XFormsContextStack.BindingContext(null, null, Collections.singletonList(documentWrapper.wrap(outputElement)), 1, "output-1", true, outputElement, null, false, null));
//
//        control1.evaluateIfNeeded(pipelineContext);
//
//        assertEquals("", control1.getExternalValue(pipelineContext));
//    }
}